File: merge_sort.py
"""
Recursive
"""


def merge_sort(l):
    length = len(l)

    if length < 2:
        return l

    left = l[: length // 2]
    right = l[length // 2:]

    return merge(merge_sort(left), merge_sort(right))


def merge(a, b):
    a_len = len(a)
    b_len = len(b)

    if a_len == 0 and b_len > 0:
        return b
    if b_len == 0 and a_len > 0:
        return a

    ah, *at = a
    bh, *bt = b

    if ah < bh:
        return [ah] + merge(at, b)
    else:
        return [bh] + merge(a, bt)

File: merge-sort/merge_sort.ml
(* Recursive *)

let rec merge_sort = function
  | []
  | [_] as l -> l
  | l ->
      let a, b = halve l in
        merge (merge_sort a, merge_sort b)
and halve = function
  | []
  | [_] as ta -> ta, []
  | h::t ->
      let ta, tb = halve t in
        h::tb, ta
and merge = function
  | l, []
  | [], l -> l
  | ha::ta, hb::tb ->
      if ha <= hb then
        ha :: merge (ta, hb::tb)
      else
        hb :: merge (ha::ta, tb)
;;
